src/allmydata/webish.py

```diff
@@ -65,18 +77,24 @@ def requestReceived(self, command, path, version):
             self.path, argstring = x
             self.args = parse_qs(argstring, 1)
 
-        if self.method == 'POST':
+        if self.method == b'POST':
             # We use FieldStorage here because it performs better than
             # cgi.parse_multipart(self.content, pdict) which is what
             # twisted.web.http.Request uses.
-            self.fields = FieldStorage(
-                self.content,
-                {
-                    name.lower(): value[-1]
-                    for (name, value)
-                    in self.requestHeaders.getAllRawHeaders()
-                },
-                environ={'REQUEST_METHOD': 'POST'})
+
+            headers = {
+                ensure_str(name.lower()): ensure_str(value[-1])
+                for (name, value)
+                in self.requestHeaders.getAllRawHeaders()
+            }
+
+            if 'content-length' not in headers:
+                # Python 3's cgi module would really, really like us to set
+                # Content-Length. This seems likely to shoot performance in
+                # the foot.
+                headers['content-length'] = len(self.content.getvalue())
```

Comment:
Not sure what to do about this one. I investigated hooking into `getContentFile` with a class that tracks bytes written, but in the `test_webish` tests (at least) it appears we're not using `TahoeLAFSSite`.

Reply:
Should that be `str(len(...))`?

For speed, maybe something like:

```
self.content.seek(0, 2) # or whatever the symbolic constant is instead of magic
length = self.content.tell()
self.content.seek(0, 0)
```

And I guess it does the last line at the end already.

Reply:
> Should that be `str(len(...))`

Oh, probably. I had decided not to care since `cgi` is gonna reconvert back to `int`, and it's highly unlikely anyone else is going to look at `self.fields.headers['content-length']` downstream of this method and not also want it to be an `int`. But I suppose it's theoretically possible, and anyway `str` is more correct, yes. :)

Reply:
> And I guess it does the last line at the end already.

Yeah, but we should hand it off to `FieldStorage` in a consistent state.

Reply:
Cleaned up in bd402ce1f411cc97b7dd81a210ff897049e2ce67, thanks.